src/build_vehicle_gui.cpp

```diff
@@ -1010,6 +1304,67 @@ void DisplayVehicleSortDropDown(Window *w, VehicleType vehicle_type, int selecte
 	ShowDropDownMenu(w, _engine_sort_listing[vehicle_type], selected, button, 0, hidden_mask);
 }
 
+bool IsEnginePassFilter(const VehicleFilter& filter, const Engine* engine) {
+	if (!filter.show_hidden_engines && engine->IsHidden(_local_company)) return false;
+	if (!IsEngineBuildable(engine->index, VEH_TRAIN, _local_company)) return false;
+
+	const RailFilter& rail_filter = filter.rail;
+	uint hasPower = engine->GetPower() > 0;
+	bool hasCapacity = engine->CanCarryCargo();
+
+	bool is_locomotive = !hasCapacity && hasPower;
+	bool is_motor_wagon = hasCapacity && hasPower;
+	bool is_wagon = hasCapacity && !hasPower;
+
+	if (!is_locomotive && !is_motor_wagon && !is_wagon) {
+		return false;
+	}
+	if (is_locomotive && !rail_filter.locomotive) {
+		return false;
+	}
+	if (is_motor_wagon && !rail_filter.motor_wagon) {
+		return false;
+	}
+	if (is_wagon && !rail_filter.wagon) {
+		return false;
+	}
+	if (!is_locomotive && !CargoFilterFunc(&engine->index, filter.cargo.type)) {
+		return false;
+	}
+	if (is_wagon) {
+		// speed in km/h
+		uint speed = engine->GetDisplayMaxSpeed();
+
+		if (rail_filter.train.speed > speed) return false;
+	}
+	if (is_locomotive && rail_filter.train.weight > 0 && rail_filter.train.speed > 0) {
+		// speed in km/h
+		uint speed = engine->GetDisplayMaxSpeed();
+		// power in kW
+		uint power = engine->GetPower() * 1000 / 735;
+		// weight in T
+		uint weight = rail_filter.train.weight + engine->GetDisplayWeight();
+		// force in kN
+		uint tractive_effort = engine->GetDisplayMaxTractiveEffort();
+
+		uint length = rail_filter.train.length + (VEHICLE_LENGTH - engine->u.rail.shorten_factor);
+
+		if (rail_filter.train.speed > speed) return false;
+
+		uint min_tractive_effort = weight * 35 / 1000;
+		if (rail_filter.slope_used && rail_filter.train.length > 0) {
+			min_tractive_effort += _settings_game.vehicle.train_slope_steepness * weight / 10 * min(rail_filter.train.length, rail_filter.slope_length) / rail_filter.train.length;
```

Comment:
This definitely needs a comment explaining the purpose and why it's correct/works as intended.